src/app/policies/views/PolicyTypeListView.vue

```diff
@@ -0,0 +1,161 @@
+<template>
+  <DataSource
+    v-slot="{ data: me }: MeSource"
+    src="/me"
+  >
+    <RouteView
+      v-if="me"
+      v-slot="{ route, t }"
+      name="policy-list-view"
+      :params="{
+        page: 1,
+        size: me.pageSize,
+        mesh: '',
+        policyPath: '',
+        policy: '',
+      }"
+    >
+      <AppView>
+        <template #title>
+          <h2>
+            <RouteTitle
+              :title="t('policies.routes.items.title')"
+            />
+          </h2>
+        </template>
+        <DataSource
+          v-slot="{ data: meshInsight }: MeshInsightSource"
+          :src="`/mesh-insights/${route.params.mesh}`"
+        >
+          <DataSource
+            v-slot="{ data }: PolicyTypeCollectionSource"
+            :src="`/*/policy-types`"
+          >
+            <div
+              class="policy-list-content"
+            >
+              <KCard
+                class="policy-type-list"
+                data-testid="policy-type-list"
+              >
+                <!-- block on policy types but not meshInsight -->
+                <DataLoader
+                  :data="[data]"
+                >
+                  <!-- this DataCollection will provide us with an empty state -->
+                  <DataCollection
+                    v-slot="{ items }"
+                    :items="data!.policies"
+                  >
+                    <template
+                      v-for="legacy in [typeof meshInsight?.policies === 'undefined' ? items : items.filter(item => {
+                        // legacy policies are those that aren't targetRef, aren't MeshGateway and are also in use
+                        return !item.isTargetRefBased && item.name !== 'MeshGateway' && (meshInsight.policies?.[item.name]?.total ?? 0) > 0
+                      })]"
+                      :key="legacy"
+                    >
```

Comment:
**Suggestion**: This is the point in the template where you should introduce a component so that you can use computed properties of data source data instead of forcing you to use these hellish `<template v-for>` constructs to create local variables out of increasingly complex inline function calls.

It’s a problem that function calls are routinely being (re-)introduced in the template, because, as I pointed out quite a few times, it forces Vue to re-render the trees as Vue can’t know whether a function needs reevaluation — it has to call them again. That’s what computed properties are for. You should use them liberally.

The current PolicyList component does a good job of that. For one, it’s great that there’s a place where loading of data happens and one where all the presentational duties of the page are carried out. For another, it’s much easier to understand what happens in a view if one doesn’t have to digest virtually _all_ its logic in one gigantic template. Finally, and most importantly, it minimizes re-renders through the use of computed properties.

Reply:
It’s a suggestion.

Reply:
I took another look at this and found I could reuse the DataCollection for the filtering here, therefore removing one instance of `<template v-for>` and reducing the gigantic(?) template from 115 lines to 107 lines. I think I can resolve this one now but let me know if not.